File: packages/expo/src/modules/clerk-expo-passkeys/ios/ClerkExpoPasskeysModule.swift
import ExpoModulesCore
import AuthenticationServices

public class ClerkExpoPasskeysModule: Module {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ClerkExpoPasskeys')` in JavaScript.
    Name("ClerkExpoPasskeys")

    AsyncFunction("create") { ( challenge: String,
                                      rpId: String,
                                      userId: String,
                                      displayName: String,
                                      promise: Promise ) in
            
            guard let challenge = Data.fromBase64Url(challenge) else {
                promise.rejectWith(passkeyError: PasskeyError.InvalidChallenge)
                return
            }
            guard let userId = Data.fromBase64Url(userId) else {
                promise.rejectWith(passkeyError: PasskeyError.InvalidUserId)
                return
            }
            
            // Passkey support came with iOS 16
            if #available(iOS 15.0, *) {
                let platformProvider = ASAuthorizationPlatformPublicKeyCredentialProvider(relyingPartyIdentifier: rpId)
                let authRequest = platformProvider.createCredentialRegistrationRequest(challenge: challenge, name: displayName, userID: userId)
                let authController = ASAuthorizationController(authorizationRequests: [authRequest])
                let passkeyDelegate = PasskeyDelegate();
                
                // Perform authorization, check for the error and parse the result
                passkeyDelegate.performAuth(for: authController, completion: { error, result in
                    if (error != nil) {
                        promise.rejectWith(passkeyError: self.convertNativeError(error: error!))
                        return
                    }
                    
                    // Check if the result object contains a valid registration result
                    if let registrationResult = result?.registrationResult {
                        // Return a NSDictionary instance with the received authorization data
                        let authResult: NSDictionary = [
                            "credentialID": registrationResult.credentialID.toBase64Url(),
                            "response": [
                                "rawAttestationObject": registrationResult.rawAttestationObject.toBase64Url(),
                                "rawClientDataJSON": registrationResult.rawClientDataJSON.toBase64Url(),
                            ]
                        ]
                        promise.resolve(authResult)
                    } else {
                        // If result didn't contain a valid registration result throw an error
                        promise.rejectWith(passkeyError: PasskeyError.RequestFailed)
                    }
                })
            } else {
                promise.rejectWith(passkeyError: PasskeyError.NotSupported)
            }
        }


        AsyncFunction("get") { (challenge: String,
                                         rpId: String,
                                         allowedCredentials: [String],
                                         promise: Promise) in
            guard let challenge = Data.fromBase64Url(challenge) else {
                promise.rejectWith(passkeyError: PasskeyError.InvalidChallenge)
                return
            }
            
            if #available(iOS 15.0, *) {
                let platformProvider = ASAuthorizationPlatformPublicKeyCredentialProvider(relyingPartyIdentifier: rpId)
                let authRequest = platformProvider.createCredentialAssertionRequest(challenge: challenge)

                // Try to parse included credentials and add it to the auth request
                do {
                    let credentialDescriptors = try parseCredentials(allowedCredentials)
                    authRequest.allowedCredentials = credentialDescriptors
                } catch let error as PasskeyError {
                    promise.rejectWith(passkeyError: error)
                    return
                }
                let authController = ASAuthorizationController(authorizationRequests: [authRequest])
                let passkeyDelegate = PasskeyDelegate();
                passkeyDelegate.performAuth(for: authController, completion: { error, result in
                    if (error != nil) {
                        promise.rejectWith(passkeyError: self.convertNativeError(error: error!))
                        return
                    }
                    
                    // Check if the result object contains a valid authentication result
                    if let authenticationResult = result?.authenticationResult {
                        let authResult: NSDictionary = [
                            "credentialID": authenticationResult.credentialID.toBase64Url(),
                            "userID": authenticationResult.userID.toBase64Url(),
                            "response": [
                                "rawAuthenticatorData": authenticationResult.rawAuthenticatorData.toBase64Url(),
                                "rawClientDataJSON": authenticationResult.rawClientDataJSON.toBase64Url(),
                                "signature": authenticationResult.signature.toBase64Url(),
                            ]
                        ]
                        promise.resolve(authResult)
                    } else {
                        promise.rejectWith(passkeyError: PasskeyError.RequestFailed)
                    }
                })
            } else {
                promise.rejectWith(passkeyError: PasskeyError.NotSupported)
            }
        }
    }


    func convertNativeError(error: Error) -> PasskeyError {
        let errorCode = (error as NSError).code
        switch errorCode {
        case 1001:
            return PasskeyError.Cancelled
        case 1004:
            return PasskeyError.RequestFailed
        case 4004:
            return PasskeyError.NotConfigured
        default:
            return PasskeyError.UnknownError
        }
    }


    @available(iOS 15, *)
    func parseCredentials(_ credentials: [String]) throws -> [ASAuthorizationPlatformPublicKeyCredentialDescriptor] {
        guard credentials.count > 0 else { return [] }

        var formattedCredentials: [Data] = []
        for credential in credentials {
            guard let data = Data.fromBase64Url(credential) else {
                throw PasskeyError.InvalidChallenge
            }
            formattedCredentials.append(data)
        }

        return formattedCredentials.map { ASAuthorizationPlatformPublicKeyCredentialDescriptor(credentialID: $0) }
    }
        

    // Defines event names that the module can send to JavaScript.
    // Events("onChange")

    // Defines a JavaScript function that always returns a Promise and whose native code
    // is by default dispatched on the different thread than the JavaScript runtime runs on.
    // AsyncFunction("setValueAsync") { (value: String) in
    //   // Send an event to JavaScript.
    //   self.sendEvent("onChange", [
    //     "value": value
    //   ])
    // }

    // Enables the module to be used as a native view. Definition components that are accepted as part of the
    // view definition: Prop, Events.
  //   View(ClerkExpoPasskeysView.self) {
  //     // Defines a setter for the `name` prop.
  //     Prop("name") { (view: ClerkExpoPasskeysView, prop: String) in
  //       print(prop)
  //     }
  //   }
  // }
}

File: packages/expo/modules/clerk-expo-passkeys/android/src/main/java/expo/modules/clerkexpopasskeys/ClerkExpoPasskeysModule.kt
package expo.modules.clerkexpopasskeys

import AuthenticationResponseJSON
import PublicKeyCredentialCreationOptions
import PublicKeyCredentialRequestOptions
import RegistrationResponseJSON
import androidx.credentials.CreatePublicKeyCredentialRequest
import androidx.credentials.CredentialManager
import androidx.credentials.GetCredentialRequest
import androidx.credentials.GetPublicKeyCredentialOption
import androidx.credentials.exceptions.CreateCredentialCancellationException
import androidx.credentials.exceptions.CreateCredentialException
import androidx.credentials.exceptions.CreateCredentialInterruptedException
import androidx.credentials.exceptions.CreateCredentialProviderConfigurationException
import androidx.credentials.exceptions.CreateCredentialUnknownException
import androidx.credentials.exceptions.CreateCredentialUnsupportedException
import androidx.credentials.exceptions.GetCredentialCancellationException
import androidx.credentials.exceptions.GetCredentialException
import androidx.credentials.exceptions.GetCredentialInterruptedException
import androidx.credentials.exceptions.GetCredentialProviderConfigurationException
import androidx.credentials.exceptions.GetCredentialUnknownException
import androidx.credentials.exceptions.GetCredentialUnsupportedException
import androidx.credentials.exceptions.NoCredentialException
import androidx.credentials.exceptions.publickeycredential.CreatePublicKeyCredentialDomException
import androidx.credentials.exceptions.publickeycredential.GetPublicKeyCredentialDomException
import com.google.gson.Gson
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch


class ClerkExpoPasskeysModule : Module() {
  private val mainScope = CoroutineScope(Dispatchers.Default)
  
  override fun definition() = ModuleDefinition {
      Name("ClerkExpoPasskeys")

      Function("isSupported") {
          val minApiLevelPasskeys = 28
          val currentApiLevel = android.os.Build.VERSION.SDK_INT
          return@Function currentApiLevel >= minApiLevelPasskeys
      }

      Function("isAutoFillAvailable") {
          false
      }

        AsyncFunction("get") { request: String, promise: Promise ->
            val credentialManager =
                CredentialManager.create(appContext.reactContext?.applicationContext!!)
            val getCredentialRequest =
                GetCredentialRequest(listOf(GetPublicKeyCredentialOption(request)))
    
            mainScope.launch {
                try {
                    val result = appContext.activityProvider?.currentActivity?.let {
                        credentialManager.getCredential(it, getCredentialRequest)
                    }
                    val response =
                     result?.credential?.data?.getString("androidx.credentials.BUNDLE_KEY_AUTHENTICATION_RESPONSE_JSON")
                    val authenticationResponse =
                        Gson().fromJson(response, AuthenticationResponseJSON::class.java)
                    promise.resolve(authenticationResponse)
                } catch (e: GetCredentialException) {
                    promise.reject("Passkey Get", e.message, e)
                }
            }
        }

      

      AsyncFunction("create") { request: String, promise: Promise ->
          val credentialManager =
              CredentialManager.create(appContext.reactContext?.applicationContext!!)
          val createPublicKeyCredentialRequest = CreatePublicKeyCredentialRequest(request)


          mainScope.launch {
              try {
                  val result = appContext.activityProvider?.currentActivity?.let {
                      credentialManager.createCredential(it, createPublicKeyCredentialRequest)
                  }
                  val response =
                      result?.data?.getString("androidx.credentials.BUNDLE_KEY_REGISTRATION_RESPONSE_JSON")
                  val createCredentialResponse =
                      Gson().fromJson(response, RegistrationResponseJSON::class.java)
                  promise.resolve(createCredentialResponse)
              } catch (e: CreateCredentialException) {
                  promise.reject("Passkey Create", e.message, e)
              }
          }
      }

    // Defines event names that the module can send to JavaScript.
    Events("onChange")

    // Defines a JavaScript synchronous function that runs the native code on the JavaScript thread.
    Function("hello") {
      "Hello world! 👋"
    }

    // Defines a JavaScript function that always returns a Promise and whose native code
    // is by default dispatched on the different thread than the JavaScript runtime runs on.
    AsyncFunction("setValueAsync") { value: String ->
      // Send an event to JavaScript.
      sendEvent("onChange", mapOf(
        "value" to value
      ))
    }

    // Enables the module to be used as a native view. Definition components that are accepted as part of
    // the view definition: Prop, Events.
    View(ClerkExpoPasskeysView::class) {
      // Defines a setter for the `name` prop.
      Prop("name") { view: ClerkExpoPasskeysView, prop: String ->
        println(prop)
      }
    }
  }
}
